### PR TITLE
itest: fix off-by-one mining while waiting for CSV-encumbered outputs

### PR DESCRIPTION
**Pull request by matheusd**
Many instances where the itests mine `defaultCSV` blocks (to wait for a sweep of a CSV-encumbered output) have an off-by-one error. This causes flaky tests when the CI slows downs during mining.

This fixes those tests by introducing a debug function to slowly mine blocks (thus causing the buggy tests to show themselves) then adjusting the number of mined blocks as needed.

### CODE REVIEW

File: lntest/harness.go
Comment (Roasbeef) [inline lntest/harness.go:1388]: So the plan is to remove this method in a final commit? I think we should also remove it all together before this PR is merged, unless we think it's useful for other things.

File: lntest/itest/lnd_multi-hop_htlc_local_timeout_test.go
Comment (Roasbeef) [inline lntest/itest/lnd_multi-hop_htlc_local_timeout_test.go]: Hmm, IMO all the `-1` everywhere (but not sometimes) makes the ordering in the tests at times hard to reason about. I think this could do with a comment explaining that it's a `-1` here since we mined the transaction confirming the commitment above (this is the case right?).

File: lntest/itest/lnd_multi-hop_htlc_local_timeout_test.go
Comment (Roasbeef) [inline lntest/itest/lnd_multi-hop_htlc_local_timeout_test.go]: How did the blocks being mined very quickly mask this?

File: lntest/itest/lnd_multi-hop_htlc_remote_chain_claim_test.go
Comment (Roasbeef) [inline lntest/itest/lnd_multi-hop_htlc_remote_chain_claim_test.go]: Same comment here re adding a commit about the `-1`.

File: lntest/itest/lnd_multi-hop_local_force_close_on_chain_htlc_timeout_test.go
Comment (Roasbeef) [inline lntest/itest/lnd_multi-hop_local_force_close_on_chain_htlc_timeout_test.go]: Same thing here that we already mined a block earlier confirming it? I don't see blocks being mined earlier though.

File: lntest/harness.go
Comment (matheusd) [reply lntest/harness.go:1388]: Personally, I'd leave it since it's sometimes useful to quickly test out replacing the regular mining functions for this, but I'll defer to whatever you prefer. I can remove it once this is cleared for merging.

File: lntest/itest/lnd_multi-hop_htlc_local_timeout_test.go
Comment (matheusd) [reply lntest/itest/lnd_multi-hop_htlc_local_timeout_test.go]: Due to the interaction with the previous `mineBlocks`, when fast mining, the sweeper (which uses a 2s timeout in itests) is not reliably triggered every single block.

So when you fast mine, you actually pass the first eligible block where some sweeping tx would be broadcast. `slowMineBlocks` ensures the sweeper is triggered at every block.

File: lntest/itest/lnd_multi-hop_htlc_local_timeout_test.go
Comment (matheusd) [reply lntest/itest/lnd_multi-hop_htlc_local_timeout_test.go]: It's been a while, but IIRC the main underlying issue is that the contract resolver sends the output for sweeping on the block immediately before the sweeping tx could be included in a block. So if the commitment output is eligible for inclusion in a block at `[current +] defaultCSV`, the sweeping tx is broadcast at block `[current +] defaultCSV-1` so if you want to reliably wait for that tx in the mempool, you need to do the subtraction.

When fast mining, the blocks come in faster then the contractcourt+sweeper can send the tx to the mempool, so it doesn't make a difference (modulo flakes). When slow mining, the contractcourt+sweeper have enough time to sweep any possible outputs at every block, thus you can reliably trigger the issue.

**Comment by halseth** [comment]
ping @matheusd

**Comment by matheusd** [comment]
Sorry for the delay. I added a commit with the review fixes. Once they are OK'd, I'll drop the intermediate commits and the slow generate functions.

File: lntest/harness.go
Comment (guggero) [reply lntest/harness.go:1423]: If we don't use the methods, the linter will probably complain about them being unused.
On the other hand, maybe it would make sense to actually run the slow mining code in the itests if that surfaces the problem more reliably. Perhaps 3 seconds is a bit extreme and we can use a shorter interval that still surfaces the problem.
How much extra time does this add to the tests? When we can actually run them in parallel with other tests (see #4655) a few extra seconds might not matter much if we have more reliable tests.

Comment (matheusd) [reply lntest/harness.go:1423]: However many blocks are mined (for the serial case) if you do this on every block. Don't know exact count for lnd but probably upwards of 1000+ blocks.

Comment (guggero) [reply lntest/harness.go:1423]: I meant just leaving in the slow mining added by 35e7b7f638e56f36e56c13721a93ad60a00a79a0 and reducing the values in 3feddc0b41345a0033b85ce316f52e2a29f600b7 to 1 second instead of 3. But I guess this still adds up to ~40 seconds (`finalCltvDelta`) to the `local_force_close_on-chain_htlc_timeout` test.
So probably better remove the calls to `slowMineBlocks`. I'm fine with leaving the functions themselves, they'll just need a `//nolint` comment (probably).

File: lntest/itest/lnd_multi-hop_htlc_local_timeout_test.go
Comment (guggero) [inline lntest/itest/lnd_multi-hop_htlc_local_timeout_test.go:227]: nit: double space at start of comment.

**Review by guggero — COMMENTED** [review]
I can't speak to the correctness of the CSV values as I'm not too familiar with that part of the code base. But it would be great to get this PR in.
I'm okay with leaving the `slowMineBlocks` function(s) in the code and only removing the calls to it.

**Review by halseth — CHANGES_REQUESTED** [review]
ACK! The slow generate methods can be removed, and commits squashed :)

**Comment by matheusd** [comment]
Squashed and removed intermediate code

**Comment by guggero** [comment]
Needs a rebase which will also fix the error whitelist problem in the itests.

**Comment by matheusd** [comment]
Rebased. Seems the outstanding failure is due to a different flake.

**Review by guggero — APPROVED** [review]
Thanks for rebasing, LGTM!

There are still flakes left, that's correct. But we're hunting them one by one now :beetle:

File: lntest/itest/lnd_multi-hop_htlc_remote_chain_claim_test.go
Comment (halseth) [inline lntest/itest/lnd_multi-hop_htlc_remote_chain_claim_test.go:107]: nit: extra newline

File: lntest/itest/lnd_multi-hop_htlc_remote_chain_claim_test.go
Comment (halseth) [inline lntest/itest/lnd_multi-hop_htlc_remote_chain_claim_test.go:112]: nit: typo

**Review by halseth — APPROVED** [review]
Thanks for this PR, less flakes always brings joy going into the weekend 🕺

File: lntest/itest/lnd_multi-hop_htlc_remote_chain_claim_test.go
Comment (guggero) [reply lntest/itest/lnd_multi-hop_htlc_remote_chain_claim_test.go:107]: Added a commit to fix this in #4655.

File: lntest/itest/lnd_multi-hop_htlc_remote_chain_claim_test.go
Comment (guggero) [reply lntest/itest/lnd_multi-hop_htlc_remote_chain_claim_test.go:112]: Added a commit to fix this in #4655.

**Pull request merged**
